### PR TITLE
Add native integer keyword recommenders

### PR DESCRIPTION
**Pull request by cston**
Relates to https://github.com/dotnet/roslyn/issues/38821 (native ints feature)

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    internal abstract class AbstractNativeIntegerKeywordRecommender : IKeywordRecommender<CSharpSyntaxContext>
+    {
+        protected abstract RecommendedKeyword Keyword { get; }
+
+        private bool IsValidContext(CSharpSyntaxContext context)
+        {
+            if (context.IsStatementContext ||
+                context.IsGlobalStatementContext ||
+                context.IsPossibleTupleContext ||
+                context.IsPatternContext ||
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs:23]: 1. can you add a pattern test and a tuple test.  
2.  also a test that it doesn't appear here `enum e : $$`
3. what aobut here `nint n = new $$`
4. not after `as`
5. yes after `is`
6. after `fixed`?
7. in lambda parameter
8. `static operator implicit $$`
9. `<see cref="$$"/>`
10. after `ref`?
11. after const? (unless it's not legal there)
12. after `stackalloc`?
13. return type of a member in a class?

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.KeywordRecommenders
+{
+    internal abstract class AbstractNativeIntegerKeywordRecommender : IKeywordRecommender<CSharpSyntaxContext>
+    {
+        protected abstract RecommendedKeyword Keyword { get; }
+
+        private bool IsValidContext(CSharpSyntaxContext context)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs:18]: static?

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/NintKeywordRecommender.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.Completion.Providers;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.KeywordRecommenders
+{
+    internal sealed class NintKeywordRecommender : AbstractNativeIntegerKeywordRecommender
+    {
+        protected override RecommendedKeyword Keyword => new RecommendedKeyword("nint");
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/NintKeywordRecommender.cs:11]: `nameof(nint)`?

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs:23]: need a test for a normal expression location.  like `var v = 1 + $$`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs [hunk 4]
```diff
@@ -1932,6 +1932,27 @@ static void verify(CSharpCompilation comp)
             }
         }
 
+        [WorkItem(42975, "https://github.com/dotnet/roslyn/issues/42975")]
+        [Fact]
+        public void AliasName_02()
+        {
+            var source =
+@"using N = nint;
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:1940]: Why does this allowed, but `using I = int;` isn't? Seems inconsistent.

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs:1940]: Agreed. [https://github.com/dotnet/roslyn/issues/42975](https://github.com/dotnet/roslyn/issues/42975)

---
In reply to: [401738499](https://github.com/dotnet/roslyn/pull/42976#discussion_r401738499) [](ancestors = 401738499)

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/NintKeywordRecommender.cs (hunk 3 above)
Comment (cston) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/NintKeywordRecommender.cs:11]: Yes, we can make that change when the feature is included in the build.

---
In reply to: [401372085](https://github.com/dotnet/roslyn/pull/42976#discussion_r401372085) [](ancestors = 401372085)

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs (hunk 1 above)
Comment (cston) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs:23]: Great suggestions, thanks.

---
In reply to: [401371992](https://github.com/dotnet/roslyn/pull/42976#discussion_r401371992) [](ancestors = 401371992)

Comment (tannergooding) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs:23]: > also a test that it doesn't appear here enum e : $$

I thought LDM agreed this was fine provided it didn't require unreasonable changes? I can't actually find the notes for that meeting however...

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs:23]: > I thought LDM agreed this was fine provided it didn't require unreasonable changes? I can't actually find the notes for that meeting however...

If so, then add test that it's legal there :)

Comment (tannergooding) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs:23]: Ah, here we go: https://github.com/dotnet/csharplang/blob/3b22aa0b47cc2ab6352e01f73f976fdbbf9d9b2d/meetings/2019/LDM-2019-10-23.md

> For use as an enum underlying type, we think nint/nuint should be allowed as long as it's not too much implementation work. The legal values would be the same as the valid constants for the underlying.

Comment (cston) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs:23]: I'll include the `enum` context when native integers are supported as underlying types in the compiler.

---
In reply to: [401782265](https://github.com/dotnet/roslyn/pull/42976#discussion_r401782265) [](ancestors = 401782265)

**Pull request merged**